src/platform/imx8ulp/lib/dai.c

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright 2021 NXP
+//
+// Author: Zhang Peng <peng.zhang_8@nxp.com>
+
+#include <sof/common.h>
+#include <sof/drivers/edma.h>
+#include <sof/drivers/sai.h>
+#include <sof/lib/dai.h>
+#include <sof/lib/memory.h>
+#include <sof/sof.h>
+#include <sof/spinlock.h>
+#include <ipc/dai.h>
+#include <ipc/stream.h>
+
+static SHARED_DATA struct dai sai[] = {
+{
+	.index = 5,
+	.plat_data = {
+		.base = SAI_5_BASE,
+		.fifo[SOF_IPC_STREAM_PLAYBACK] = {
+			.offset		= SAI_5_BASE + REG_SAI_TDR0,
+			/* use depth to model the HW FIFO size:
+			 * each channel includes a 16 x 32 bit FIFO
+			 * that can be accessed using Transmit or
+			 * Receive Data Registers
+			 */
+			.depth		= 16,  /* in 4 bytes words */
+			.watermark      = 8,
+			.handshake	= EDMA_HANDSHAKE(EDMA2_8ULP_CHAN0_IRQ,
+							 EDMA2_8ULP_CHAN0),
+		},
+		.fifo[SOF_IPC_STREAM_CAPTURE] = {
+			.offset		= SAI_5_BASE + REG_SAI_RDR2,
+			.depth		= 16,  /* in 4 bytes words */
+			.watermark      = 8,
+			.handshake	= EDMA_HANDSHAKE(EDMA2_8ULP_CHAN1_IRQ,
+							 EDMA2_8ULP_CHAN1),
+		},
+		.dmamux_rx_num = DMAMUX2_SAI5_RX_NUM,
+		.dmamux_tx_num = DMAMUX2_SAI5_TX_NUM,
+	},
```

Comment:
Please move the changes with dmamux_rx_num in a separate patch together with the changes that add DMAMUX2_SAI5_TX_NUM. 

This is important to be in a separate patch because it touches core structures. In this huge PR it can be easily missed by Intel people.

They will be most likely to be interested in core changes and they will go easy on IMX changes.

Reply:
Later edit: not sure if this even compiles. We need to modify `struct dai` right?

Reply:
Yes. Need to modify dai first.

Reply:
@TangleZ as discussed internally, please do not bother adding this stuff in the SOF core (dmamux, srcid). This is can be easily sorted out in the EDMA driver with an ifdef IMX8ULP. 
I think you were on a right path by trying to make this generic but this touches kernel core and if we want to have something with quality and working we can make this compromise now.